### DISCUSSION TITLE
Missing DisplayAsync Razor Helper Extension

### DISCUSSION THREAD

**Discussion by mrpavlitzky** [OrchardCMS/OrchardCore#12214] [Q&A]
Hi Community,

I came across a following issue:

I want to use the Orchard.DisplayAsync() method, but I get the following error:

`CS1929	'IOrchardHelper' does not contain a definition for 'DisplayAsync' and the best extension method overload 'OrchardRazorHelperExtensions.DisplayAsync(IOrchardDisplayHelper, ContentItem, string, string, IUpdateModel)' requires a receiver of type 'IOrchardDisplayHelper'`

The argument is of type ContentItem and the Helper is injected via `@inject OrchardCore.IOrchardHelper Orchard`.

Any way around this?

**Comment by jtkech** [comment]
Our custom RazorPage already has a `IOrchardDisplayHelper Orchard` property that you can use directly, nothing to inject.

For this you need a `_ViewImports` including

    @inherits OrchardCore.DisplayManagement.Razor.RazorPage<TModel>

**Reply by mrpavlitzky** [reply]
I did what you suggested, now I'm sitting on a bunch of errors:

![Capture](https://user-images.githubusercontent.com/77694733/185318172-8803e0c1-f00b-420a-8fa1-81c10dfc0c74.JPG)

**Reply by hishamco** [reply]
Check your SDK, also check if Razor is support for views.

Don't forgot to reference the required project references

**Reply by jtkech** [reply]
@Cinnam00n 

Based on your errors it's about Razor @pages (under the Pages folder) not razor views (under the Views folder)

So in your `_ViewImports` if under the Pages folder, in place of

    @inherits OrchardCore.DisplayManagement.Razor.RazorPage<TModel>

Use this instead

    @inherits OrchardCore.DisplayManagement.RazorPages.Page

**Reply by mrpavlitzky** [reply]
@jtkech 

When I switch to the second code snippet you provided I get the error about `@RenderBody()` and `@ViewData[]` not existing in current context which is pretty confusing as I see that your custom Razor Page inherits from the base `Mvc.RazorPages.Page` so it should have those methods availalable. 

The error about hiding inherited members still persist:

```'Pages_BlogPost.Orchard' hides inherited member 'Page.Orchard'. Use the new keyword if hiding was intended.	OrchardCore.Cms.Web (net6.0) ```

**Comment by jtkech** [comment]
`@RenderBody()`  is part of the mvc `Razor.RazorPage` but not part of the mvc `RazorPages.Page`.

Note: In the context of Orchard using the theming engine, we don't use `@RenderBody()` but `@await RenderBodyAsync()`, normally in a `Layout.cshtml` under the `/Views` folder of the selected theme.

See the `OC.Demo` module in the source code, there are example of @pages under the module `/Pages` folder. In a given page just tried to use `@ViewData`, it was suggested by intellisense.

> hides inherited member 'Page.Orchard'

Looks like you are still injecting `@inject OrchardCore.IOrchardHelper Orchard`

**Reply by mrpavlitzky** [reply]
> Note: In the context of Orchard using the theming engine, we don't use `@RenderBody()` but `@await RenderBodyAsync()`, normally in a `Layout.cshtml` under the `/Views` folder of the selected theme.

I'm working in a decoupled architecture with only Pages for now (no theming engine).

> See the `OC.Demo` module in the source code, there are example of @pages under the module `/Pages` folder. In a given page just tried to use `@ViewData`, it was suggested by intellisense.

I've studied it carefully and I can see it working, but in my case adding `@inherits OrchardCore.DisplayManagement.RazorPages.Page` breaks the `ViewData' and `RenderBody()` functions. Without the inherit statement, the functions are referenced correctly. 


> Looks like you are still injecting `@inject OrchardCore.IOrchardHelper Orchard`

I cleared all other Orchard namespace references and the error disappeared.

**Reply by mrpavlitzky** [reply]
Also, `ViewData[]` is available in any of the .cshtml files. It's missing only from the `_Layout`

**Comment by mrpavlitzky** [comment]
Okay, I'm getting somewhere. It seems that we could reformulate the original question with: 

**Is it possible to render shapes in a decoupled architecture (without Views and Templates)**?

Why?

I originally started with decoupled site, because of the lowest learning curve. The project I work on is not massively complicated (basically a blog with some a-little-more-complex-than-usual tagging system, comments and possibility of creating a list of liked posts stored in LocalStorage/Cookies). But now there's [Disqus module](https://github.com/agriffard/Disqus.OrchardCore) I want to implement, which apparently is to be rendered as a shape. 

So the actual question is:

**Is there a way to render the Disqus Part without using shapes system?**

**Reply by ns8482e** [reply]
May be you need to have different discussion questions with more detail on what you are really trying achieve.

short answer to your question is
> Is there a way to render the Disqus Part without using shapes system?

No, ContentPart shapes and drivers are part of shape system

However you can control which shape  you want to render using  DisplayAsync
